In [6]:
# Step 4. Importing fundamental libraries
import pandas as pd 
import numpy as np
import os 
import requests
import json
from datetime import datetime

In [13]:
#Checking for working directory

print("Current working dir:", os.getcwd())
print("Parent folder contents:", os.listdir(".."))

Current working dir: /Users/javieraquezada/Desktop/Citi-Bike-/Tasks
Parent folder contents: ['Tasks', 'Requirements Texts', '.DS_Store', 'README.md', '.git', 'Data']


In [16]:
# Creating a list with all files in the folder using a list compehension
import os

folderpath = os.path.join("..", "Data")  # go up one level, then into Data1
filepaths = [os.path.join(folderpath, name) for name in os.listdir(folderpath)]

print(filepaths)

['../Data/JC-202203-citibike-tripdata.csv', '../Data/.DS_Store', '../Data/JC-202201-citibike-tripdata.csv', '../Data/JC-202111-citibike-tripdata.csv', '../Data/JC-202107-citibike-tripdata.csv', '../Data/JC-202209-citibike-tripdata.csv', '../Data/JC-202109-citibike-tripdata.csv', '../Data/JC-202211-citibike-tripdata.csv', '../Data/JC-202207-citbike-tripdata.csv', '../Data/JC-202205-citibike-tripdata.csv', '../Data/JC-202112-citibike-tripdata.csv', '../Data/JC-202202-citibike-tripdata.csv', '../Data/JC-202208-citibike-tripdata.csv', '../Data/JC-202110-citibike-tripdata.csv', '../Data/JC-202206-citibike-tripdata.csv', '../Data/JC-202210-citibike-tripdata.csv', '../Data/JC-202108-citibike-tripdata.csv', '../Data/JC-202204-citibike-tripdata.csv']


In [17]:
filepaths

['../Data/JC-202203-citibike-tripdata.csv',
 '../Data/.DS_Store',
 '../Data/JC-202201-citibike-tripdata.csv',
 '../Data/JC-202111-citibike-tripdata.csv',
 '../Data/JC-202107-citibike-tripdata.csv',
 '../Data/JC-202209-citibike-tripdata.csv',
 '../Data/JC-202109-citibike-tripdata.csv',
 '../Data/JC-202211-citibike-tripdata.csv',
 '../Data/JC-202207-citbike-tripdata.csv',
 '../Data/JC-202205-citibike-tripdata.csv',
 '../Data/JC-202112-citibike-tripdata.csv',
 '../Data/JC-202202-citibike-tripdata.csv',
 '../Data/JC-202208-citibike-tripdata.csv',
 '../Data/JC-202110-citibike-tripdata.csv',
 '../Data/JC-202206-citibike-tripdata.csv',
 '../Data/JC-202210-citibike-tripdata.csv',
 '../Data/JC-202108-citibike-tripdata.csv',
 '../Data/JC-202204-citibike-tripdata.csv']

In [ ]:
# Step 4. Read and concatenate all files simultaneously

df = pd.concat((pd.read_csv(f) for f in filepaths), ignore_index = True)

UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 3131: invalid start byte

In [ ]:
#Second try of reading and concatenate all files simultaneously.Telling pandas to try a different encoding while reading files.
import pandas as pd
import glob

filepaths = glob.glob('../Data/*.csv')

df = pd.concat((pd.read_csv(f, encoding="latin1") for f in filepaths), ignore_index=True)

In [ ]:
# Files listing.
import os

data_folder = '../Data'
print(os.listdir(data_folder))

['JC-202203-citibike-tripdata.csv', '.DS_Store', 'JC-202201-citibike-tripdata.csv', 'JC-202111-citibike-tripdata.csv', 'JC-202107-citibike-tripdata.csv', 'JC-202209-citibike-tripdata.csv', 'JC-202109-citibike-tripdata.csv', 'JC-202211-citibike-tripdata.csv', 'JC-202207-citbike-tripdata.csv', 'JC-202205-citibike-tripdata.csv', 'JC-202112-citibike-tripdata.csv', 'JC-202202-citibike-tripdata.csv', 'JC-202208-citibike-tripdata.csv', 'JC-202110-citibike-tripdata.csv', 'JC-202206-citibike-tripdata.csv', 'JC-202210-citibike-tripdata.csv', 'JC-202108-citibike-tripdata.csv', 'JC-202204-citibike-tripdata.csv']


In [21]:
df

ride_id  rideable_type           started_at  \
0        3255D3E3F33CDC45   classic_bike  2022-03-18 15:38:17   
1        17FA5604A37338F9  electric_bike  2022-03-04 16:44:48   
2        7DEC9ADDB8D6BBE1  electric_bike  2022-03-13 17:44:32   
3        9D69F74EEF231A2E   classic_bike  2022-03-13 15:33:47   
4        C84AE4A9D78A6347   classic_bike  2022-03-11 12:21:18   
...                   ...            ...                  ...   
1317451  13C1C453508F048A   classic_bike  2022-04-24 12:52:09   
1317452  B7E454B8886D7826  electric_bike  2022-04-30 11:45:56   
1317453  20D6A9E7C8291984   classic_bike  2022-04-28 14:28:34   
1317454  2AEB863A4F683CAE  electric_bike  2022-04-09 17:37:14   
1317455  AB2C885E7051B2D4   classic_bike  2022-04-02 15:54:12   

                    ended_at                            start_station_name  \
0        2022-03-18 15:45:34        Mama Johnson Field - 4 St & Jackson St   
1        2022-03-04 16:50:45                         Baldwin at Montgomery   
2        2022-03-13 17:54:44                         Baldwin at Montgomery   
3        2022-03-13 15:41:22                         Baldwin at Montgomery   
4        2022-03-11 12:33:24                         Baldwin at Montgomery   
...                      ...                                           ...   
1317451  2022-04-24 13:06:24  South Waterfront Walkway - Sinatra Dr & 1 St   
1317452  2022-04-30 11:54:30  South Waterfront Walkway - Sinatra Dr & 1 St   
1317453  2022-04-28 14:41:35  South Waterfront Walkway - Sinatra Dr & 1 St   
1317454  2022-04-09 17:46:45  South Waterfront Walkway - Sinatra Dr & 1 St   
1317455  2022-04-02 16:02:53  South Waterfront Walkway - Sinatra Dr & 1 St   

        start_station_id                              end_station_name  \
0                  HB404  South Waterfront Walkway - Sinatra Dr & 1 St   
1                  JC020                                 Grove St PATH   
2                  JC020                                 Grove St PATH   
3                  JC020                                 Grove St PATH   
4                  JC020                                 Grove St PATH   
...                  ...                                           ...   
1317451            HB103                 9 St HBLR - Jackson St & 8 St   
1317452            HB103                          12 St & Sinatra Dr N   
1317453            HB103                         Bloomfield St & 15 St   
1317454            HB103                         Bloomfield St & 15 St   
1317455            HB103                             Clinton St & 7 St   

        end_station_id  start_lat  start_lng    end_lat    end_lng  \
0                HB103  40.743140 -74.040041  40.736982 -74.027781   
1                JC005  40.723659 -74.064194  40.719586 -74.043117   
2                JC005  40.723659 -74.064194  40.719586 -74.043117   
3                JC005  40.723659 -74.064194  40.719586 -74.043117   
4                JC005  40.723659 -74.064194  40.719586 -74.043117   
...                ...        ...        ...        ...        ...   
1317451          HB305  40.736982 -74.027781  40.747907 -74.038412   
1317452          HB201  40.736982 -74.027781  40.750604 -74.024020   
1317453          HB203  40.736982 -74.027781  40.754530 -74.026580   
1317454          HB203  40.736982 -74.027781  40.754530 -74.026580   
1317455          HB303  40.736982 -74.027781  40.745420 -74.033320   

        member_casual  
0              casual  
1              member  
2              member  
3              member  
4              member  
...               ...  
1317451        member  
1317452        member  
1317453        member  
1317454        casual  
1317455        casual  

[1317456 rows x 13 columns]

In [22]:
df.head()


ride_id  rideable_type           started_at             ended_at  \
0  3255D3E3F33CDC45   classic_bike  2022-03-18 15:38:17  2022-03-18 15:45:34   
1  17FA5604A37338F9  electric_bike  2022-03-04 16:44:48  2022-03-04 16:50:45   
2  7DEC9ADDB8D6BBE1  electric_bike  2022-03-13 17:44:32  2022-03-13 17:54:44   
3  9D69F74EEF231A2E   classic_bike  2022-03-13 15:33:47  2022-03-13 15:41:22   
4  C84AE4A9D78A6347   classic_bike  2022-03-11 12:21:18  2022-03-11 12:33:24   

                       start_station_name start_station_id  \
0  Mama Johnson Field - 4 St & Jackson St            HB404   
1                   Baldwin at Montgomery            JC020   
2                   Baldwin at Montgomery            JC020   
3                   Baldwin at Montgomery            JC020   
4                   Baldwin at Montgomery            JC020   

                               end_station_name end_station_id  start_lat  \
0  South Waterfront Walkway - Sinatra Dr & 1 St          HB103  40.743140   
1                                 Grove St PATH          JC005  40.723659   
2                                 Grove St PATH          JC005  40.723659   
3                                 Grove St PATH          JC005  40.723659   
4                                 Grove St PATH          JC005  40.723659   

   start_lng    end_lat    end_lng member_casual  
0 -74.040041  40.736982 -74.027781        casual  
1 -74.064194  40.719586 -74.043117        member  
2 -74.064194  40.719586 -74.043117        member  
3 -74.064194  40.719586 -74.043117        member  
4 -74.064194  40.719586 -74.043117        member

In [23]:
df.tail()


ride_id  rideable_type           started_at  \
1317451  13C1C453508F048A   classic_bike  2022-04-24 12:52:09   
1317452  B7E454B8886D7826  electric_bike  2022-04-30 11:45:56   
1317453  20D6A9E7C8291984   classic_bike  2022-04-28 14:28:34   
1317454  2AEB863A4F683CAE  electric_bike  2022-04-09 17:37:14   
1317455  AB2C885E7051B2D4   classic_bike  2022-04-02 15:54:12   

                    ended_at                            start_station_name  \
1317451  2022-04-24 13:06:24  South Waterfront Walkway - Sinatra Dr & 1 St   
1317452  2022-04-30 11:54:30  South Waterfront Walkway - Sinatra Dr & 1 St   
1317453  2022-04-28 14:41:35  South Waterfront Walkway - Sinatra Dr & 1 St   
1317454  2022-04-09 17:46:45  South Waterfront Walkway - Sinatra Dr & 1 St   
1317455  2022-04-02 16:02:53  South Waterfront Walkway - Sinatra Dr & 1 St   

        start_station_id               end_station_name end_station_id  \
1317451            HB103  9 St HBLR - Jackson St & 8 St          HB305   
1317452            HB103           12 St & Sinatra Dr N          HB201   
1317453            HB103          Bloomfield St & 15 St          HB203   
1317454            HB103          Bloomfield St & 15 St          HB203   
1317455            HB103              Clinton St & 7 St          HB303   

         start_lat  start_lng    end_lat    end_lng member_casual  
1317451  40.736982 -74.027781  40.747907 -74.038412        member  
1317452  40.736982 -74.027781  40.750604 -74.024020        member  
1317453  40.736982 -74.027781  40.754530 -74.026580        member  
1317454  40.736982 -74.027781  40.754530 -74.026580        casual  
1317455  40.736982 -74.027781  40.745420 -74.033320        casual

Step 5. Code Explanation.

In [24]:
# Used this code for the following reasons:
# 1st : Loads pandas library for data analysis
# 2nd : Loads the glob library, which is used for pattern-matching file paths.
# 3rd :  filepaths = glob.glob('../Data/*.csv')
#	glob.glob() searches for all files matching the pattern.
#	Here: "../Data/*.csv" means:
#	Go up one directory (.. from Tasks).
#	Enter the Data folder.
#	Match all files ending in .csv
# 4th : This is the key line:
    #A.	(pd.read_csv(f, encoding="latin1") for f in filepaths)
    #A generator that reads each CSV file into a temporary pandas DataFrame.
    #encoding="latin1" tells pandas how to interpret the raw bytes in the file as text.
    # That’s why I got a UnicodeDecodeError earlier.
    # Using latin1 (ISO-8859-1) is more permissive and can decode any byte 0–255, so it avoids that crash.
    #B.	pd.concat(..., ignore_index=True)
    #Concatenates all the DataFrames from the generator into one big DataFrame.
    #gnore_index=True resets the row index, so you get a clean 0…N index instead of keeping each file’s original row numbers.

Basically what we did with this code we took all the csv files and asked to read all the files including the weird symbols so it doesnt get confused. And we combined it in a bigger Data frame.

In [25]:
##Checking all the csv have the same columnd


filepaths = glob.glob('../Data/*.csv')

columns_per_file = {}
for f in filepaths:
    df_temp = pd.read_csv(f, encoding="latin1", nrows=5)  # only peek at first 5 rows
    columns_per_file[f] = df_temp.columns.tolist()

for f, cols in columns_per_file.items():
    print(f"\n{f}:")
    print(cols)


../Data/JC-202203-citibike-tripdata.csv:
['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id', 'start_lat', 'start_lng', 'end_lat', 'end_lng', 'member_casual']

../Data/JC-202201-citibike-tripdata.csv:
['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id', 'start_lat', 'start_lng', 'end_lat', 'end_lng', 'member_casual']

../Data/JC-202111-citibike-tripdata.csv:
['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id', 'start_lat', 'start_lng', 'end_lat', 'end_lng', 'member_casual']

../Data/JC-202107-citibike-tripdata.csv:
['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id', 'start_lat', 'start_lng', 'end_lat', 'end_lng', 'member_casual']

../Data/JC-202209-citibike-trip

In [ ]:
#because this is a classic trap when concatenating many CSVs from different months. Sometimes CitiBike (or any dataset provider)
#adds or removes columns between months. If we don’t handle that, pandas will either fail or silently misalign data.

In [ ]:
#"Always match by the column title. If something is missing on one sheet, just leave it blank.""

filepaths = glob.glob('../Data/*.csv')

dfs = []
for f in filepaths:
    df_temp = pd.read_csv(f, encoding="latin1")
    dfs.append(df_temp)

# Concatenate all, aligning by column names
df = pd.concat(dfs, ignore_index=True, sort=True)

In [27]:
df

end_lat    end_lng end_station_id  \
0        40.736982 -74.027781          HB103   
1        40.719586 -74.043117          JC005   
2        40.719586 -74.043117          JC005   
3        40.719586 -74.043117          JC005   
4        40.719586 -74.043117          JC005   
...            ...        ...            ...   
1317451  40.747907 -74.038412          HB305   
1317452  40.750604 -74.024020          HB201   
1317453  40.754530 -74.026580          HB203   
1317454  40.754530 -74.026580          HB203   
1317455  40.745420 -74.033320          HB303   

                                     end_station_name             ended_at  \
0        South Waterfront Walkway - Sinatra Dr & 1 St  2022-03-18 15:45:34   
1                                       Grove St PATH  2022-03-04 16:50:45   
2                                       Grove St PATH  2022-03-13 17:54:44   
3                                       Grove St PATH  2022-03-13 15:41:22   
4                                       Grove St PATH  2022-03-11 12:33:24   
...                                               ...                  ...   
1317451                 9 St HBLR - Jackson St & 8 St  2022-04-24 13:06:24   
1317452                          12 St & Sinatra Dr N  2022-04-30 11:54:30   
1317453                         Bloomfield St & 15 St  2022-04-28 14:41:35   
1317454                         Bloomfield St & 15 St  2022-04-09 17:46:45   
1317455                             Clinton St & 7 St  2022-04-02 16:02:53   

        member_casual           ride_id  rideable_type  start_lat  start_lng  \
0              casual  3255D3E3F33CDC45   classic_bike  40.743140 -74.040041   
1              member  17FA5604A37338F9  electric_bike  40.723659 -74.064194   
2              member  7DEC9ADDB8D6BBE1  electric_bike  40.723659 -74.064194   
3              member  9D69F74EEF231A2E   classic_bike  40.723659 -74.064194   
4              member  C84AE4A9D78A6347   classic_bike  40.723659 -74.064194   
...               ...               ...            ...        ...        ...   
1317451        member  13C1C453508F048A   classic_bike  40.736982 -74.027781   
1317452        member  B7E454B8886D7826  electric_bike  40.736982 -74.027781   
1317453        member  20D6A9E7C8291984   classic_bike  40.736982 -74.027781   
1317454        casual  2AEB863A4F683CAE  electric_bike  40.736982 -74.027781   
1317455        casual  AB2C885E7051B2D4   classic_bike  40.736982 -74.027781   

        start_station_id                            start_station_name  \
0                  HB404        Mama Johnson Field - 4 St & Jackson St   
1                  JC020                         Baldwin at Montgomery   
2                  JC020                         Baldwin at Montgomery   
3                  JC020                         Baldwin at Montgomery   
4                  JC020                         Baldwin at Montgomery   
...                  ...                                           ...   
1317451            HB103  South Waterfront Walkway - Sinatra Dr & 1 St   
1317452            HB103  South Waterfront Walkway - Sinatra Dr & 1 St   
1317453            HB103  South Waterfront Walkway - Sinatra Dr & 1 St   
1317454            HB103  South Waterfront Walkway - Sinatra Dr & 1 St   
1317455            HB103  South Waterfront Walkway - Sinatra Dr & 1 St   

                  started_at  
0        2022-03-18 15:38:17  
1        2022-03-04 16:44:48  
2        2022-03-13 17:44:32  
3        2022-03-13 15:33:47  
4        2022-03-11 12:21:18  
...                      ...  
1317451  2022-04-24 12:52:09  
1317452  2022-04-30 11:45:56  
1317453  2022-04-28 14:28:34  
1317454  2022-04-09 17:37:14  
1317455  2022-04-02 15:54:12  

[1317456 rows x 13 columns]

Consistency checks


In [28]:
# Checking for duplicated rows
dupes = df.duplicated().sum()
print(f"Number of duplicate rows: {dupes}")

Number of duplicate rows: 0


In [29]:
# Checking trip duration consistency
if "tripduration" in df.columns:
    print("Min trip duration:", df["tripduration"].min())
    print("Max trip duration:", df["tripduration"].max())

In [30]:
# Checking stations and id names
if "start_station_name" in df.columns:
    print("Unique start stations:", df["start_station_name"].nunique())

if "end_station_name" in df.columns:
    print("Unique end stations:", df["end_station_name"].nunique())

Unique start stations: 87
Unique end stations: 371


In [31]:
#As stations should be similar in both ends we are going to investigate further.
# finding stations that are only on end but never in start
start_stations = set(df["start_station_name"].dropna().unique())
end_stations   = set(df["end_station_name"].dropna().unique())

only_in_end = end_stations - start_stations
only_in_start = start_stations - end_stations

print("Stations only in END:", len(only_in_end))
print("Stations only in START:", len(only_in_start))

Stations only in END: 285
Stations only in START: 1


In [ ]:
#Peeking at them
print("Examples only in END:", list(only_in_end)[:20])
print("Examples only in START:", list(only_in_start)[:20])

Examples only in END: ['William St & Pine St', 'W 34 St & 11 Ave', 'West End Ave & W 107 St', 'MacDougal St & Washington Sq', '11 Ave & W 27 St', 'W 34 St &\\tHudson Blvd E', 'Cliff St & Fulton St', 'Broad St & Bridge St', 'South St & Gouverneur Ln', 'Shop Morgan', 'E 9 St & 5 Ave', 'W 54 St & 11 Ave', 'E 20 St & 2 Ave', '6 Ave & Walker St', 'Van Sinderen Ave & Truxton St', 'Suffolk St & Stanton St', 'W 37 St & 5 Ave', 'Central Park North & Adam Clayton Powell Blvd', 'Duane St & Greenwich St', 'Wadsworth Ave & W 175 St']
Examples only in START: ['E 20 St & FDR Drive']


What this might mean
- Data quality issue
- Some trips were logged with an end_station_name that isn’t valid (like "NULL", "0", "test station").
- Encoding or naming inconsistencies
- Example: "Grove St PATH" vs "Grove St. PATH " (extra space, dot, casing).
- Pandas treats these as different stations, even though they’re the same.
- System changes
- CitiBike sometimes adds/removes stations, but 285 missing starts is too many to be real.

In [33]:
# Checking for weird cases
print(df["start_station_name"].value_counts().head(10))
print(df["end_station_name"].value_counts().head(10))

start_station_name
Grove St PATH                                   60512
South Waterfront Walkway - Sinatra Dr & 1 St    53386
Hoboken Terminal - Hudson St & Hudson Pl        47530
Hoboken Terminal - River St & Hudson Pl         47184
Newport Pkwy                                    34752
Hamilton Park                                   31816
Newport PATH                                    31388
Hoboken Ave at Monmouth St                      29144
Marin Light Rail                                29053
14 St Ferry - 14 St & Shipyard Ln               27291
Name: count, dtype: int64
end_station_name
Grove St PATH                                   62925
South Waterfront Walkway - Sinatra Dr & 1 St    53934
Hoboken Terminal - Hudson St & Hudson Pl        47702
Hoboken Terminal - River St & Hudson Pl         46253
Newport Pkwy                                    34553
Hamilton Park                                   32123
Newport PATH                                    31344
Hoboken Ave at Monmo

In [34]:
# Checking stations only in end
print("Stations only in END (sample):")
print(list(only_in_end)[:50])  # just to peek

Stations only in END (sample):
['William St & Pine St', 'W 34 St & 11 Ave', 'West End Ave & W 107 St', 'MacDougal St & Washington Sq', '11 Ave & W 27 St', 'W 34 St &\\tHudson Blvd E', 'Cliff St & Fulton St', 'Broad St & Bridge St', 'South St & Gouverneur Ln', 'Shop Morgan', 'E 9 St & 5 Ave', 'W 54 St & 11 Ave', 'E 20 St & 2 Ave', '6 Ave & Walker St', 'Van Sinderen Ave & Truxton St', 'Suffolk St & Stanton St', 'W 37 St & 5 Ave', 'Central Park North & Adam Clayton Powell Blvd', 'Duane St & Greenwich St', 'Wadsworth Ave & W 175 St', 'Murray St\\t& West St', 'Adam Clayton Powell Blvd & W 132 St', 'W 50 St & 10 Ave', 'Liberty St & Nassau St', 'W 104 St & Amsterdam Ave', 'W 116 St & Broadway', 'Faile St & Garrison Ave', 'Broadway & Morris St', 'W 100 St & Manhattan Ave', 'Cabrini Blvd & W 177 St', 'W 33 St & 10 Ave', 'Fulton St & William St', 'Clinton St & Cherry St', 'Amsterdam Ave & W 79 St', 'W 17 St & 8 Ave', 'Stanton St & Chrystie St', 'Soissons Landing', 'E 16 St & 5 Ave', 'E 39 St & 2

In [35]:
#checking weird placeholders
print(df["end_station_name"].value_counts().head(20))

end_station_name
Grove St PATH                                   62925
South Waterfront Walkway - Sinatra Dr & 1 St    53934
Hoboken Terminal - Hudson St & Hudson Pl        47702
Hoboken Terminal - River St & Hudson Pl         46253
Newport Pkwy                                    34553
Hamilton Park                                   32123
Newport PATH                                    31344
Hoboken Ave at Monmouth St                      30160
Marin Light Rail                                28661
14 St Ferry - 14 St & Shipyard Ln               28027
Columbus Dr at Exchange Pl                      26717
City Hall - Washington St & 1 St                26582
11 St & Washington St                           24913
Harborside                                      24632
Washington St                                   23738
City Hall                                       23042
12 St & Sinatra Dr N                            22795
8 St & Washington St                            22576
Liberty Lig

In [ ]:
# 2. Likely causes of mismatch
#	1.	Whitespace and tabs
#	'W 34 St &\\tHudson Blvd E' has a tab \t in the string → counts as a different station.
#	2.	Case differences
#	'William St & Pine St' vs 'william st & pine st' → pandas sees them as different.
#	3.	Trailing/leading spaces
#	'Broad St & Bridge St ' vs 'Broad St & Bridge St' → counts as different.
#	4.	Actual missing stations
#	Some stations may have only existed in certain months → the one “only in START” might be real.

In [39]:
# Normalizing stations names
def clean_station_name(s):
    if pd.isna(s):
        return s
    return s.strip().replace("\t", " ").lower()

df["start_station_name"] = df["start_station_name"].apply(clean_station_name)
df["end_station_name"] = df["end_station_name"].apply(clean_station_name)


In [40]:
# checking mismatches
start_stations = set(df["start_station_name"].dropna().unique())
end_stations = set(df["end_station_name"].dropna().unique())

only_in_end = end_stations - start_stations
only_in_start = start_stations - end_stations

print("Stations only in END:", len(only_in_end))
print("Stations only in START:", len(only_in_start))

Stations only in END: 285
Stations only in START: 1


In [38]:
print("Unique start stations:", df["start_station_name"].nunique())
print("Unique end stations:", df["end_station_name"].nunique())

Unique start stations: 87
Unique end stations: 371


In [ ]:
# nothing changes so we arrived to the conclusion that is not a cleaning problem, it is more like how data is shown and behaves.
# probably some stations are used only as destinations.
# maybe the data set can be incomplete for certain months.
# probably some stations were temporarely open.

In [ ]:
# Flaging stations for further analysis if necesary later.
start_stations = set(df["start_station_name"].dropna().unique())
end_stations = set(df["end_station_name"].dropna().unique())

station_types = {}
for s in start_stations.union(end_stations):
    if s in start_stations and s in end_stations:
        station_types[s] = "both"
    elif s in start_stations:
        station_types[s] = "start_only"
    else:
        station_types[s] = "end_only"



In [44]:
# checking how many stations fall in each category
from collections import Counter
Counter(station_types.values())

Counter({'end_only': 285, 'both': 86, 'start_only': 1})

- 285 stations are end_only : trips end here but never start.
- 86 stations are both : appear as both start and end.
- 1 station is start_only : trips start here but never end.

For flow analysis or mapping trips, 285 stations are always destinations, 1 is always a start, and 86 are both.

In [48]:
# Adding an extra column to show flag

In [ ]:
# Adding columns to the DataFrame
df["start_station_type"] = df["start_station_name"].map(station_types)
df["end_station_type"]   = df["end_station_name"].map(station_types)


In [47]:
# Quick check
df[["start_station_name", "start_station_type", "end_station_name", "end_station_type"]].head()

start_station_name start_station_type  \
0  mama johnson field - 4 st & jackson st               both   
1                   baldwin at montgomery               both   
2                   baldwin at montgomery               both   
3                   baldwin at montgomery               both   
4                   baldwin at montgomery               both   

                               end_station_name end_station_type  
0  south waterfront walkway - sinatra dr & 1 st             both  
1                                 grove st path             both  
2                                 grove st path             both  
3                                 grove st path             both  
4                                 grove st path             both

In [ ]:
# Step 6. Getting the NOAA's web app token.

In [ ]:
# Step 7. Obtaining weather data from New York LaGuardia's weather station for 2022 and export it to a csv file
# Saving id: GHCND:USW00014732

In [77]:
#Defining Token

Token = 'aAXrPKlmBAVmXECyDywfunHivCQBDWAt' 

In [ ]:
#Parameters for 2022

#'datasetid': 'GSOY',
#'datatypeid': 'PRCP,TAVG,TMAX,TMIN,TSUN',
#'stationid': 'GHCND:USW00014732',
#'startdate': '2022-01-01',
#'enddate': '2022-12-31',
#'limit': 1000


In [78]:
#Get the API
r = requests.get('https://ncdc.noaa.gov/cdo-web/api/v2/data?datasetid=GHCND&datatypeid=PRCP,TAVG,TMAX,TMIN,TSUN&limit=1000&stationid=GHCND:USW00014732&startdate=2022-01-01&enddate=2022-12-31', 
headers={'token': Token})

In [79]:
#Load api response as Json


d = json.loads(r.text)

In [80]:
d

{'metadata': {'resultset': {'offset': 1, 'count': 1460, 'limit': 1000}},
 'results': [{'date': '2022-01-01T00:00:00',
   'datatype': 'PRCP',
   'station': 'GHCND:USW00014732',
   'attributes': ',,W,2400',
   'value': 193},
  {'date': '2022-01-01T00:00:00',
   'datatype': 'TAVG',
   'station': 'GHCND:USW00014732',
   'attributes': 'H,,S,',
   'value': 116},
  {'date': '2022-01-01T00:00:00',
   'datatype': 'TMAX',
   'station': 'GHCND:USW00014732',
   'attributes': ',,W,2400',
   'value': 139},
  {'date': '2022-01-01T00:00:00',
   'datatype': 'TMIN',
   'station': 'GHCND:USW00014732',
   'attributes': ',,W,2400',
   'value': 100},
  {'date': '2022-01-02T00:00:00',
   'datatype': 'PRCP',
   'station': 'GHCND:USW00014732',
   'attributes': ',,W,2400',
   'value': 10},
  {'date': '2022-01-02T00:00:00',
   'datatype': 'TAVG',
   'station': 'GHCND:USW00014732',
   'attributes': 'H,,S,',
   'value': 114},
  {'date': '2022-01-02T00:00:00',
   'datatype': 'TMAX',
   'station': 'GHCND:USW00014732

In [81]:
#Converting as wide format for easier analysis

# Convert results to DataFrame
df_weather = pd.DataFrame(d['results'])

# Pivot: one row per date, columns = datatypes
df_weather = df_weather.pivot(index='date', columns='datatype', values='value').reset_index()

# Convert date to datetime
df_weather['date'] = pd.to_datetime(df_weather['date'])

# Preview
print(df_weather.head())

datatype       date  PRCP  TAVG  TMAX  TMIN
0        2022-01-01   193   116   139   100
1        2022-01-02    10   114   156    39
2        2022-01-03     0    14    39   -43
3        2022-01-04     0   -27    22   -60
4        2022-01-05    61    32    89     0


In [82]:
# Change of scale values due to NOAA sometimes storing temperatures as tenths of °C and precipitation as hundredths of mm.
df_weather['PRCP'] = df_weather['PRCP'] / 100
df_weather['TAVG'] = df_weather['TAVG'] / 10
df_weather['TMAX'] = df_weather['TMAX'] / 10
df_weather['TMIN'] = df_weather['TMIN'] / 10

df_weather.head()

datatype       date  PRCP  TAVG  TMAX  TMIN
0        2022-01-01  1.93  11.6  13.9  10.0
1        2022-01-02  0.10  11.4  15.6   3.9
2        2022-01-03  0.00   1.4   3.9  -4.3
3        2022-01-04  0.00  -2.7   2.2  -6.0
4        2022-01-05  0.61   3.2   8.9   0.0

In [83]:
# Preparing to merge
#Checking columns

In [84]:
print(df.columns)

Index(['end_lat', 'end_lng', 'end_station_id', 'end_station_name', 'ended_at',
       'member_casual', 'ride_id', 'rideable_type', 'start_lat', 'start_lng',
       'start_station_id', 'start_station_name', 'started_at',
       'start_station_type', 'end_station_type', 'date'],
      dtype='object')


In [85]:
print(df_weather.columns)


Index(['date', 'PRCP', 'TAVG', 'TMAX', 'TMIN'], dtype='object', name='datatype')


In [86]:
#Ready to merge
# Convert started_at to datetime
df['started_at'] = pd.to_datetime(df['started_at'])

# Create a date-only column for merging
df['date'] = df['started_at'].dt.date



In [87]:
# Make sure weather date is also datetime.date
df_weather['date'] = pd.to_datetime(df_weather['date']).dt.date

# Merge CitiBike trips with weather data
df_merged = df.merge(df_weather, on='date', how='left')
df_merged.head()

end_lat    end_lng end_station_id  \
0  40.736982 -74.027781          HB103   
1  40.719586 -74.043117          JC005   
2  40.719586 -74.043117          JC005   
3  40.719586 -74.043117          JC005   
4  40.719586 -74.043117          JC005   

                               end_station_name             ended_at  \
0  south waterfront walkway - sinatra dr & 1 st  2022-03-18 15:45:34   
1                                 grove st path  2022-03-04 16:50:45   
2                                 grove st path  2022-03-13 17:54:44   
3                                 grove st path  2022-03-13 15:41:22   
4                                 grove st path  2022-03-11 12:33:24   

  member_casual           ride_id  rideable_type  start_lat  start_lng  \
0        casual  3255D3E3F33CDC45   classic_bike  40.743140 -74.040041   
1        member  17FA5604A37338F9  electric_bike  40.723659 -74.064194   
2        member  7DEC9ADDB8D6BBE1  electric_bike  40.723659 -74.064194   
3        member  9D69F74EEF231A2E   classic_bike  40.723659 -74.064194   
4        member  C84AE4A9D78A6347   classic_bike  40.723659 -74.064194   

  start_station_id                      start_station_name  \
0            HB404  mama johnson field - 4 st & jackson st   
1            JC020                   baldwin at montgomery   
2            JC020                   baldwin at montgomery   
3            JC020                   baldwin at montgomery   
4            JC020                   baldwin at montgomery   

           started_at start_station_type end_station_type        date  PRCP  \
0 2022-03-18 15:38:17               both             both  2022-03-18   0.0   
1 2022-03-04 16:44:48               both             both  2022-03-04   0.0   
2 2022-03-13 17:44:32               both             both  2022-03-13   0.0   
3 2022-03-13 15:33:47               both             both  2022-03-13   0.0   
4 2022-03-11 12:21:18               both             both  2022-03-11   0.0   

   TAVG  TMAX  TMIN  
0  13.9  22.8   9.4  
1  -1.9   3.3  -5.5  
2  -2.3   1.7  -4.9  
3  -2.3   1.7  -4.9  
4   7.2  11.7   2.2

In [88]:
#Explanation:
#	started_at is timestamp with date + time. Weather data is daily.
#	We extract only the date from started_at to align with weather dates.
#	Merge on the new date column using how='left' to keep all CitiBike trips and attach weather info.
#   This will correctly attach the precipitation, average temperature, min/max temp, to each trip.

In [89]:
#Checking if merging is correct
# Check some rows to see if weather info is attached
print(df_merged[['started_at', 'date', 'PRCP', 'TAVG', 'TMAX', 'TMIN']].head(10))

           started_at        date  PRCP  TAVG  TMAX  TMIN
0 2022-03-18 15:38:17  2022-03-18  0.00  13.9  22.8   9.4
1 2022-03-04 16:44:48  2022-03-04  0.00  -1.9   3.3  -5.5
2 2022-03-13 17:44:32  2022-03-13  0.00  -2.3   1.7  -4.9
3 2022-03-13 15:33:47  2022-03-13  0.00  -2.3   1.7  -4.9
4 2022-03-11 12:21:18  2022-03-11  0.00   7.2  11.7   2.2
5 2022-03-03 08:01:51  2022-03-03  0.08   5.6   8.9  -3.8
6 2022-03-21 22:41:49  2022-03-21  0.00  10.4  16.7   6.7
7 2022-03-20 10:42:53  2022-03-20  0.00  14.3  16.1   7.8
8 2022-03-24 17:52:22  2022-03-24  2.59   5.7   8.3   3.9
9 2022-03-19 05:29:52  2022-03-19  0.18  13.1  20.0   7.8


In [90]:
# Check for any trips without weather info
missing_weather = df_merged[df_merged['PRCP'].isna()].shape[0]
print(f"Number of trips without weather info: {missing_weather}")

Number of trips without weather info: 716172


The main reasons for this are likely:
	1.	Date format mismatch – your CitiBike date column might include time, or be in a slightly different format than the weather date.
	2.	Weather data only covers part of the year – your API request returned 1460 rows (about 4 years?), but maybe you’re merging with trips outside those dates.
	3.	Station data differences – if any trip dates fall outside the weather dataset range, they won’t match.

In [73]:
#Fixing and Checking
# Make sure both are datetime and only date part
df['date'] = pd.to_datetime(df['started_at']).dt.date
df_weather['date'] = pd.to_datetime(df_weather['date']).dt.date



In [74]:
# Check min/max dates
print(df['date'].min(), df['date'].max())
print(df_weather['date'].min(), df_weather['date'].max())

# Merge again
df_merged = df.merge(df_weather, on='date', how='left')

2021-07-01 2022-11-30
2022-01-01 2022-09-07


In [75]:
missing_weather = df_merged['PRCP'].isna().sum()
print("Number of trips without weather info:", missing_weather)

Number of trips without weather info: 716172


In [72]:
# Quick summary stats of weather columns
print(df_merged[['PRCP', 'TAVG', 'TMAX', 'TMIN']].describe())

                PRCP           TAVG           TMAX           TMIN
count  601284.000000  601284.000000  601284.000000  601284.000000
mean        0.203670      19.147844      23.813524      15.208398
std         0.498942       8.832370       9.170562       8.883835
min         0.000000      -8.000000      -5.500000     -12.100000
25%         0.000000      13.100000      18.300000       8.900000
50%         0.000000      21.900000      26.700000      17.800000
75%         0.100000      25.900000      31.100000      22.200000
max         2.950000      31.300000      36.700000      27.800000


In [91]:
# Check date range
print("Min date:", df_merged['date'].min())
print("Max date:", df_merged['date'].max())

# Check missing weather values
missing_weather = df_merged[['PRCP', 'TAVG', 'TMAX', 'TMIN']].isna().sum()
print("Missing weather data per column:\n", missing_weather)

# Count trips with any missing weather info
trips_missing_weather = df_merged[df_merged[['PRCP','TAVG','TMAX','TMIN']].isna().any(axis=1)].shape[0]
print("Number of trips without weather info:", trips_missing_weather)

Min date: 2021-07-01
Max date: 2022-11-30
Missing weather data per column:
 PRCP    716172
TAVG    716172
TMAX    716172
TMIN    716172
dtype: int64
Number of trips without weather info: 716172


 ~700k trips have missing weather info—they are from late 2022, after my weather dataset ends. 

In [92]:
df_merged.to_csv("merged_citibike_weather.csv", index=False)
print("Merged CSV exported successfully.")

Merged CSV exported successfully.


In [93]:
df_merged.to_csv("merged_citibike_weather.csv.gz", index=False, compression='gzip')